src/org/rascalmpl/uri/URIResolverRegistry.java

```diff
@@ -57,6 +59,14 @@ public class URIResolverRegistry {
 	private final Map<String, ISourceLocationWatcher> watchers = new ConcurrentHashMap<>();
 	private final Map<ISourceLocation, Set<Consumer<ISourceLocationChanged>>> watching = new ConcurrentHashMap<>();
 
+	// we allow the user to define (using -Drascal.fallbackResolver=fully.qualified.classname) a single class that will handle
+	// scheme's not statically registered. That class should implement at least one of these interfaces
+	private final AtomicReference<@Nullable ISourceLocationInput> fallbackInputResolver = new AtomicReference<>();
```

Comment:
I don't understand the motivation for the AtomicReference. All this code is run during the initialization of the class as triggered by the nested static class InstanceHolder. As far as I know, this is still a thread-safe way of initializing a singleton. Please explain, or simplify?

Reply:
Okay, it could be replaced by a volatile field. But it's due to the semantics of the loadServices.

Reply:
`loadServices` is only called from the constructor right? Which is in the InstanceHolder. It does not even have to be `volatile` in my mind.

Reply:
It's about multiple threads accessing it. And we have the 'reload' feature.

Reply:
I am looking for a more specific answer. Which multiple threads would be able to access these fields simultaneously? (leave reinitialize out of this discussion for clarity).

`reinitialize` is definitely a topic for thread-safety, but at the moment not called from anywhere and this could also be handled by making that method `synchronized`

Reply:
Volatile & atomic reference (which is just a volatile field with some extra features) are not about locking, they are about visibility & caching (but they can be used to avoid locking).

I'll remove the atomic reference to explain it a bit more clearly:

```
private volatile ISourceLocationInput fallbackInputResolver;
```

Note this field cannot be final, since it's not initialized by the constructor. (it's initialized during the constructor, so that's good news, but it's not final). If it were final, we wouldn't need a volatile field.

So, it only goes wrong after a `reinitialize` and a `synchronized` would not help. If thread A has used the `URIResolverRegistry` it's allowed to keep a cpu local cache of the field values. If thread B triggers a reload (and has done a `System.setProperty`) the value of the `fallbackInputResolver` would change. If that field was *not* volatile, you have no guarantees when thread A would see the new value of `fallbackInputResolver`.

Reply:
ok. thanks!

> So, it only goes wrong after a reinitialize and a synchronized would not help

I'm not interested in rewriting `reinitialize` so let's stick with the `volatile` solution for now. I don't think we actually have code that triggers this behavior and I also don't think the `volatile` will make the code much slower. I could be wrong, but let's see. I just want to try and keep the code as simple as possible.

Perhaps we should see how/if `reinitialize` can go again. I'll study the Eclipse code and the library resolver which promoted this design decision.

Reply:
yes, that looks better to me. 

your analysis makes me suspect that `reinitialize` is perhaps broken in more than one way. Let's add an issue for that.

Reply:
ok, fields are now volatile.